File: src/lib/worker/wiki/docs.ts
/**
 * Docs-wiki agent.
 *
 * Reads `documents` rows in status='pending' OR re-indexed (where the page
 * hash changed) and:
 *   1. Chunks the body_text into ~1500-char chunks with 200-char overlap.
 *   2. Stores chunks in `doc_chunks`.
 *   3. Asks the doc-writer LLM to write a wiki page summarizing the document
 *      with structure: ## What it is · ## Key points · ## When to reference.
 *   4. Upserts a `wiki_pages` row with kind='docs', slug derived from filename.
 *
 * Embeddings are optional (only populated when OPENAI_API_KEY is set).
 * For v0.1 we skip the embedder call to keep dependency surface tight.
 */
import { generateText } from "ai";
import { catalogPool } from "../../catalog/db";
import { upsertWikiPage } from "../../catalog/wiki";
import { audit } from "../../catalog/queries";
import { docWriterModel } from "../openrouter";
import { withRules } from "../rules";

const ACTOR = "wiki-agent:docs";
const CHUNK_SIZE = 1500;
const CHUNK_OVERLAP = 200;

const SYSTEM_PROMPT = `You are Loom's documentation summarizer. Given the raw
text of a single document, write a concise wiki page with this exact structure:

  ## What it is
  One paragraph: what this document covers and who it's for.

  ## Key points
  3 to 8 bullet points. Quote a phrase verbatim if it captures a definition,
  rule, or threshold the team should remember. Bold any KPI or metric name.

  ## When to reference
  2 or 3 bullets describing when an analyst should pull this doc into a
  question (e.g. "any question about Q3 quality targets").

  ## Related (optional)
  Bullet list of [[tables/<slug>]] or [[code/<slug>]] cross-references IF
  the document explicitly mentions table names or repos that look like Loom
  identifiers. Skip the section entirely if no obvious matches.

Write nothing outside these sections. No preamble.`;

export async function runDocsWikiAgent(): Promise<{ generated: number }> {
  // Pick up never-indexed docs OR docs whose content_hash differs from the
  // hash we summarized last. IS DISTINCT FROM treats NULL on either side as
  // "different", so a re-upload (last_indexed_hash still NULL after status
  // got flipped externally) also re-indexes.
  const pending = await catalogPool.query<{
    id: number;
    display_name: string;
    body_text: string;
    content_hash: string;
    mime: string;
    size_bytes: string;
  }>(
    `SELECT id, display_name, body_text, content_hash, mime, size_bytes::text
       FROM documents
      WHERE status = 'pending'
         OR (status = 'indexed' AND content_hash IS DISTINCT FROM last_indexed_hash)
      ORDER BY uploaded_at ASC
      LIMIT 10`,
  );

  let generated = 0;
  for (const doc of pending.rows) {
    try {
      await indexOneDoc(doc);
      generated++;
    } catch (e) {
      await catalogPool.query(`UPDATE documents SET status = 'failed' WHERE id = $1`, [doc.id]);
      await audit(ACTOR, "wiki:doc_index_failed", String(doc.id), { err: String(e) });
    }
  }
  return { generated };
}

async function indexOneDoc(doc: {
  id: number;
  display_name: string;
  body_text: string;
  content_hash: string;
}): Promise<void> {
  // 1. Chunk + (re)write chunks
  const chunks = chunkText(doc.body_text, CHUNK_SIZE, CHUNK_OVERLAP);
  await catalogPool.query(`DELETE FROM doc_chunks WHERE document_id = $1`, [doc.id]);
  for (let i = 0; i < chunks.length; i++) {
    await catalogPool.query(
      `INSERT INTO doc_chunks (document_id, ord, text) VALUES ($1, $2, $3)`,
      [doc.id, i, chunks[i]],
    );
  }

  // 2. Ask the doc-writer for a structured summary. We pass the first ~3
  // chunks plus the last chunk to cover bookends — for very long docs this
  // beats sending only the front matter.
  const sample = composeSample(chunks);
  const result = await generateText({
    model: docWriterModel(),
    system: withRules(SYSTEM_PROMPT, "wiki"),
    prompt: `# Document: ${doc.display_name}\n\n${sample}`,
    maxRetries: 1,
  });

  // 3. Upsert the wiki page.
  const slug = slugify(doc.display_name);
  const summary = firstLineOf(result.text) ?? `Summary of ${doc.display_name}`;
  await upsertWikiPage(ACTOR, {
    kind: "docs",
    slug,
    title: doc.display_name,
    summary,
    body_md: result.text.trim(),
    source_ref: { document_id: doc.id, content_hash: doc.content_hash },
  });

  // 4. Mark document indexed. Stamp last_indexed_hash with the content_hash
  // we just summarized; the next tick's WHERE clause uses this to skip the
  // LLM call when nothing has changed.
  await catalogPool.query(
    `UPDATE documents
        SET status = 'indexed',
            indexed_at = NOW(),
            last_indexed_hash = $2
      WHERE id = $1`,
    [doc.id, doc.content_hash],
  );
}

function chunkText(text: string, size: number, overlap: number): string[] {
  if (text.length <= size) return [text];
  const out: string[] = [];
  let i = 0;
  while (i < text.length) {
    const end = Math.min(i + size, text.length);
    out.push(text.slice(i, end));
    if (end === text.length) break;
    i = end - overlap;
  }
  return out;
}

function composeSample(chunks: string[]): string {
  if (chunks.length <= 4) return chunks.join("\n\n---\n\n");
  const head = chunks.slice(0, 3);
  const tail = chunks.slice(-1);
  return [...head, "…", ...tail].join("\n\n---\n\n");
}

function slugify(name: string): string {
  const base = name.replace(/\.[^.]+$/, "");
  const s = base
    .toLowerCase()
    .replace(/\s+/g, "-")
    .replace(/[^a-z0-9-]/g, "")
    .replace(/-+/g, "-")
    .replace(/^-|-$/g, "");
  return s || `doc-${Date.now()}`;
}

function firstLineOf(md: string): string | null {
  for (const line of md.split("\n")) {
    const t = line.trim();
    if (!t || t.startsWith("#")) continue;
    return t.slice(0, 200);
  }
  return null;
}

File: backend/app/workers/wiki/docs.py
"""Docs-wiki agent.

Picks up never-indexed documents (or those whose content_hash changed since
the last indexing pass), chunks them, asks the doc-writer LLM for a structured
summary, and upserts a ``wiki_pages`` row with kind='docs'. Hash-gated end to
end via ``upsert_wiki_page`` + ``documents.last_indexed_hash``.
"""
from __future__ import annotations

import logging
import re
import time
from typing import Any

from psycopg.rows import dict_row

from ...audit import audit
from ...catalog.wiki import upsert_wiki_page
from ...db import get_conn
from ...llm import async_client, chat_model_id
from ..rules import with_rules

log = logging.getLogger(__name__)

ACTOR = "wiki-agent:docs"
CHUNK_SIZE = 1500
CHUNK_OVERLAP = 200

SYSTEM_PROMPT = """You are Loom's documentation summarizer. Given the raw
text of a single document, write a concise wiki page with this exact structure:

  ## What it is
  One paragraph: what this document covers and who it's for.

  ## Key points
  3 to 8 bullet points. Quote a phrase verbatim if it captures a definition,
  rule, or threshold the team should remember. Bold any KPI or metric name.

  ## When to reference
  2 or 3 bullets describing when an analyst should pull this doc into a
  question (e.g. "any question about Q3 quality targets").

  ## Related (optional)
  Bullet list of [[tables/<slug>]] or [[code/<slug>]] cross-references IF
  the document explicitly mentions table names or repos that look like Loom
  identifiers. Skip the section entirely if no obvious matches.

Write nothing outside these sections. No preamble."""


async def run_docs_wiki_agent() -> dict[str, int]:
    """Index up to 10 pending/changed documents per tick."""
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                """
                SELECT id, display_name, body_text, content_hash, mime, size_bytes::text
                  FROM documents
                 WHERE status = 'pending'
                    OR (status = 'indexed'
                        AND content_hash IS DISTINCT FROM last_indexed_hash)
                 ORDER BY uploaded_at ASC
                 LIMIT 10
                """,
            )
            pending = await cur.fetchall()

    generated = 0
    for doc in pending:
        try:
            await _index_one_doc(doc)
            generated += 1
        except Exception as e:  # noqa: BLE001
            async with get_conn() as conn:
                async with conn.cursor() as cur:
                    await cur.execute(
                        "UPDATE documents SET status = 'failed' WHERE id = %s",
                        (doc["id"],),
                    )
            await audit(ACTOR, "wiki:doc_index_failed", str(doc["id"]), {"err": str(e)})
            log.exception("docs.index_failed id=%s", doc["id"])
    return {"generated": generated}


async def _index_one_doc(doc: dict[str, Any]) -> None:
    doc_id = int(doc["id"])
    display_name: str = doc["display_name"]
    body_text: str = doc["body_text"] or ""
    content_hash: str = doc["content_hash"]

    # 1. Chunk + (re)write chunks.
    chunks = _chunk_text(body_text, CHUNK_SIZE, CHUNK_OVERLAP)
    async with get_conn() as conn:
        async with conn.cursor() as cur:
            await cur.execute("DELETE FROM doc_chunks WHERE document_id = %s", (doc_id,))
            for i, chunk in enumerate(chunks):
                await cur.execute(
                    "INSERT INTO doc_chunks (document_id, ord, text) VALUES (%s, %s, %s)",
                    (doc_id, i, chunk),
                )

    # 2. Ask the doc-writer for a structured summary.
    sample = _compose_sample(chunks)
    client = async_client()
    resp = await client.chat.completions.create(
        model=chat_model_id(),
        messages=[
            {"role": "system", "content": with_rules(SYSTEM_PROMPT, "wiki")},
            {"role": "user", "content": f"# Document: {display_name}\n\n{sample}"},
        ],
    )
    text = (resp.choices[0].message.content or "").strip() if resp.choices else ""

    # 3. Upsert the wiki page.
    slug = _slugify(display_name)
    summary = _first_line_of(text) or f"Summary of {display_name}"
    await upsert_wiki_page(
        ACTOR,
        kind="docs",
        slug=slug,
        title=display_name,
        summary=summary,
        body_md=text,
        source_ref={"document_id": doc_id, "content_hash": content_hash},
    )

    # 4. Mark document indexed; stamp last_indexed_hash so unchanged docs skip
    #    the LLM next tick.
    async with get_conn() as conn:
        async with conn.cursor() as cur:
            await cur.execute(
                """
                UPDATE documents
                   SET status = 'indexed',
                       indexed_at = NOW(),
                       last_indexed_hash = %s
                 WHERE id = %s
                """,
                (content_hash, doc_id),
            )


def _chunk_text(text: str, size: int, overlap: int) -> list[str]:
    if len(text) <= size:
        return [text] if text else []
    out: list[str] = []
    i = 0
    n = len(text)
    while i < n:
        end = min(i + size, n)
        out.append(text[i:end])
        if end == n:
            break
        i = end - overlap
    return out


def _compose_sample(chunks: list[str]) -> str:
    if len(chunks) <= 4:
        return "\n\n---\n\n".join(chunks)
    head = chunks[:3]
    tail = chunks[-1:]
    return "\n\n---\n\n".join([*head, "…", *tail])


_SLUG_SPACES = re.compile(r"\s+")
_SLUG_DROP = re.compile(r"[^a-z0-9-]")
_SLUG_DASHES = re.compile(r"-+")
_SLUG_EDGES = re.compile(r"^-|-$")
_EXT_RE = re.compile(r"\.[^.]+$")


def _slugify(name: str) -> str:
    base = _EXT_RE.sub("", name)
    s = base.lower()
    s = _SLUG_SPACES.sub("-", s)
    s = _SLUG_DROP.sub("", s)
    s = _SLUG_DASHES.sub("-", s)
    s = _SLUG_EDGES.sub("", s)
    return s or f"doc-{int(time.time() * 1000)}"


def _first_line_of(md: str) -> str | None:
    for line in md.split("\n"):
        t = line.strip()
        if not t or t.startswith("#"):
            continue
        return t[:200]
    return None
